File: balancer/service/logic/src/config/ServiceConfig.cc
#include "ServiceConfig.h"

#include <algorithm>
#include <set>
#include "rapidjson/stringbuffer.h"
#include "rapidjson/writer.h"
#include "tool/Util.h"

std::string ServiceConfig::json_to_map(const std::string& json)
{
	rapidjson::Document document;
	document.Parse(json.c_str());
	if(!document.IsObject())
		return err_is_not_object(json);

	if(!document.HasMember(KeyServiceConfig::service_map.c_str()))
		return err_is_not_array(KeyServiceConfig::service_map);
	const rapidjson::Value& v = document[KeyServiceConfig::service_map.c_str()];
	if(!v.IsArray())
		return err_is_not_array(KeyServiceConfig::service_map);

	std::set<std::string> check_service_name;
	const auto& array = v.GetArray();
	for(const auto& a : array)
	{
		Service service;
		if(!get_uint(a, KeyServiceConfig::service_id, service.service_id))
			return err_is_not_uint(KeyServiceConfig::service_id, service.service_name);
		if(!get_string(a, KeyServiceConfig::service_name, service.service_name))
			return err_is_not_string(KeyServiceConfig::service_name, service.service_name);

		check_service_name.insert(service.service_name);

		{
			if(!a.HasMember(KeyServiceConfig::heartbeat.c_str()))
				return err_is_not_object(KeyServiceConfig::heartbeat);
			const rapidjson::Value& cv = a[KeyServiceConfig::heartbeat.c_str()];
			if(!cv.IsObject())
				return err_is_not_object(KeyServiceConfig::heartbeat, service.service_name);

			if(!get_bool(cv, KeyServiceConfig::heartbeat_enable, service.heartbeat.heartbeat_enable))
				return err_is_not_uint(KeyServiceConfig::heartbeat_enable, service.service_name);
			if(!get_uint(cv, KeyServiceConfig::heartbeat_gap, service.heartbeat.heartbeat_gap))
				return err_is_not_uint(KeyServiceConfig::heartbeat_gap, service.service_name);
			if(!get_uint(cv, KeyServiceConfig::lose_time, service.heartbeat.lose_time))
				return err_is_not_uint(KeyServiceConfig::lose_time, service.service_name);
			if(!get_uint(cv, KeyServiceConfig::recover_time, service.heartbeat.recover_time))
				return err_is_not_uint(KeyServiceConfig::recover_time, service.service_name);
		}

		{
			if(!a.HasMember(KeyServiceConfig::depend_map.c_str()))
				return err_is_not_array(KeyServiceConfig::depend_map);
			const rapidjson::Value& cv = a[KeyServiceConfig::depend_map.c_str()];
			if(!cv.IsArray())
				return err_is_not_array(KeyServiceConfig::depend_map, service.service_name);

			const auto& brray = cv.GetArray();
			for(const auto& b : brray)
			{
				Depend depend;
				if(!get_uint(b, KeyServiceConfig::depend_service_id, depend.depend_service_id))
					return err_is_not_uint(KeyServiceConfig::depend_service_id, service.service_name);

				service.depend_map.insert(std::make_pair(depend.depend_service_id, depend));
			}

			if(service.depend_map.size() != brray.Size())
				return err_is_not_unique(KeyServiceConfig::depend_service_id, service.service_name);
		}

		{
			if(!a.HasMember(KeyServiceConfig::kv_map.c_str()))
				return err_is_not_array(KeyServiceConfig::kv_map);
			const rapidjson::Value& cv = a[KeyServiceConfig::kv_map.c_str()];
			if(!cv.IsArray())
				return err_is_not_array(KeyServiceConfig::kv_map, service.service_name);

			const auto& brray = cv.GetArray();
			for(const auto& b : brray)
			{
				KV kv;
				if(!get_string(b, KeyServiceConfig::key, kv.key))
					return err_is_not_string(KeyServiceConfig::key, service.service_name);
				if(!get_string(b, KeyServiceConfig::val, kv.val))
					return err_is_not_string(KeyServiceConfig::val, service.service_name);

				service.kv_map.insert(std::make_pair(kv.key, kv));
			}

			if(service.kv_map.size() != brray.Size())
				return err_is_not_unique(KeyServiceConfig::key, service.service_name);
		}


		{
			std::set<unsigned int> check_proc_id;
			{
				if(!a.HasMember(KeyServiceConfig::heartbeat_list.c_str()))
					return err_is_not_array(KeyServiceConfig::heartbeat_list);
				const rapidjson::Value& cv = a[KeyServiceConfig::heartbeat_list.c_str()];
				if(!cv.IsArray())
					return err_is_not_array(KeyServiceConfig::heartbeat_list, service.service_name);

				const auto& brray = cv.GetArray();
				for(const auto& b : brray)
				{
					IPInfo ip_info;
					if(!get_uint(b, KeyServiceConfig::proc_id, ip_info.proc_id))
						return err_is_not_uint(KeyServiceConfig::proc_id, service.service_name);
					if(!get_string(b, KeyServiceConfig::proc_des, ip_info.proc_des))
						return err_is_not_string(KeyServiceConfig::proc_des, service.service_name);
					if(!get_string(b, KeyServiceConfig::in_ip, ip_info.in_ip))
						return err_is_not_string(KeyServiceConfig::in_ip, service.service_name);
					if(!get_string(b, KeyServiceConfig::out_ip, ip_info.out_ip))
						return err_is_not_string(KeyServiceConfig::out_ip, service.service_name);
					if(!get_uint(b, KeyServiceConfig::port, ip_info.port))
						return err_is_not_uint(KeyServiceConfig::port, service.service_name);

					ip_info.change();
					service.heartbeat_list.push_back(ip_info);
					check_proc_id.insert(ip_info.proc_id);
				}
			}

			{
				if(!a.HasMember(KeyServiceConfig::inservice_list.c_str()))
					return err_is_not_array(KeyServiceConfig::inservice_list);
				const rapidjson::Value& cv = a[KeyServiceConfig::inservice_list.c_str()];
				if(!cv.IsArray())
					return err_is_not_array(KeyServiceConfig::inservice_list, service.service_name);

				const auto& brray = cv.GetArray();
				for(const auto& b : brray)
				{
					IPInfo ip_info;
					if(!get_uint(b, KeyServiceConfig::proc_id, ip_info.proc_id))
						return err_is_not_uint(KeyServiceConfig::proc_id, service.service_name);
					if(!get_string(b, KeyServiceConfig::proc_des, ip_info.proc_des))
						return err_is_not_string(KeyServiceConfig::proc_des, service.service_name);
					if(!get_string(b, KeyServiceConfig::in_ip, ip_info.in_ip))
						return err_is_not_string(KeyServiceConfig::in_ip, service.service_name);
					if(!get_string(b, KeyServiceConfig::out_ip, ip_info.out_ip))
						return err_is_not_string(KeyServiceConfig::out_ip, service.service_name);
					if(!get_uint(b, KeyServiceConfig::port, ip_info.port))
						return err_is_not_uint(KeyServiceConfig::port, service.service_name);

					ip_info.change();
					service.inservice_list.push_back(ip_info);
					check_proc_id.insert(ip_info.proc_id);
				}
			}

			if(check_proc_id.size() != service.heartbeat_list.size() + service.inservice_list.size())
				return err_is_not_unique(KeyServiceConfig::proc_id, service.service_name);
		}

		_service_map.insert(std::make_pair(service.service_id, service));
	}

	if(_service_map.size() != array.Size())
		return err_is_not_unique(KeyServiceConfig::service_id);
	if(check_service_name.size() != array.Size())
		return err_is_not_unique(KeyServiceConfig::service_name);

	for(auto& it_service : _service_map)
	{
		Service& service = it_service.second;
		for(auto& it_depend : service.depend_map)
		{
			Depend& depend = it_depend.second;
			auto it = _service_map.find(depend.depend_service_id);
			if(it == _service_map.end())
			{
				return err_depend_service_id_is_not_exist(depend.depend_service_id);
			}
		}
	}

	return "";
}

std::string ServiceConfig::map_to_json()
{
	rapidjson::StringBuffer sb;
	rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
	
	writer.StartObject();
	{
		writer.Key(KeyServiceConfig::service_map.c_str());
		writer.StartArray();
		{
			for(auto& it_service : _service_map)
			{
				writer.StartObject();
				{
					Service& service = it_service.second;
					writer.Key(KeyServiceConfig::service_id.c_str());
					writer.Uint(service.service_id);
					writer.Key(KeyServiceConfig::service_name.c_str());
					writer.String(service.service_name.c_str());
					
					writer.Key(KeyServiceConfig::heartbeat.c_str());
					writer.StartObject();
					{
						writer.Key(KeyServiceConfig::heartbeat_enable.c_str());
						writer.Bool(service.heartbeat.heartbeat_enable);
						writer.Key(KeyServiceConfig::heartbeat_gap.c_str());
						writer.Uint(service.heartbeat.heartbeat_gap);
						writer.Key(KeyServiceConfig::lose_time.c_str());
						writer.Uint(service.heartbeat.lose_time);
						writer.Key(KeyServiceConfig::recover_time.c_str());
						writer.Uint(service.heartbeat.recover_time);
					}	
					writer.EndObject();

					writer.Key(KeyServiceConfig::depend_map.c_str());
					writer.StartArray();
					{
						for(auto& it_kv : service.depend_map)
						{
							writer.StartObject();
							{
								Depend& depend = it_kv.second;
								writer.Key(KeyServiceConfig::depend_service_id.c_str());
								writer.Uint(depend.depend_service_id);
							}
							writer.EndObject();
						}
					}
					writer.EndArray();

					writer.Key(KeyServiceConfig::kv_map.c_str());
					writer.StartArray();
					{
						for(auto& it_kv : service.kv_map)
						{
							writer.StartObject();
							{
								KV& kv = it_kv.second;
								writer.Key(KeyServiceConfig::key.c_str());
								writer.String(kv.key.c_str());
								writer.Key(KeyServiceConfig::val.c_str());
								writer.String(kv.val.c_str());
							}
							writer.EndObject();
						}
					}
					writer.EndArray();

					writer.Key(KeyServiceConfig::heartbeat_list.c_str());
					writer.StartArray();
					{
						for(auto& ip_info : service.heartbeat_list)
						{
							writer.StartObject();
							{
								writer.Key(KeyServiceConfig::proc_id.c_str());
								writer.Uint(ip_info.proc_id);
								writer.Key(KeyServiceConfig::proc_des.c_str());
								writer.String(ip_info.proc_des.c_str());
								writer.Key(KeyServiceConfig::in_ip.c_str());
								writer.String(ip_info.in_ip.c_str());
								writer.Key(KeyServiceConfig::out_ip.c_str());
								writer.String(ip_info.out_ip.c_str());
								writer.Key(KeyServiceConfig::port.c_str());
								writer.Uint(ip_info.port);
							}
							writer.EndObject();
						}
					}
					writer.EndArray();

					writer.Key(KeyServiceConfig::inservice_list.c_str());
					writer.StartArray();
					{
						for(auto& ip_info : service.inservice_list)
						{
							writer.StartObject();
							{
								writer.Key(KeyServiceConfig::proc_id.c_str());
								writer.Uint(ip_info.proc_id);
								writer.Key(KeyServiceConfig::proc_des.c_str());
								writer.String(ip_info.proc_des.c_str());
								writer.Key(KeyServiceConfig::in_ip.c_str());
								writer.String(ip_info.in_ip.c_str());
								writer.Key(KeyServiceConfig::out_ip.c_str());
								writer.String(ip_info.out_ip.c_str());
								writer.Key(KeyServiceConfig::port.c_str());
								writer.Uint(ip_info.port);
							}
							writer.EndObject();
						}
					}
					writer.EndArray();
					
				}
				writer.EndObject();
			}
		}	
		writer.EndArray();
	}
	writer.EndObject();

	return sb.GetString();
}

bool ServiceConfig::get_string(const rapidjson::Value& v, const std::string& key, std::string& str)
{
	if(v.HasMember(key.c_str()))
	{
		const rapidjson::Value& tmp = v[key.c_str()];
		if(tmp.IsString())
		{
			str = tmp.GetString();
			return str.size() > 0;
		}
	}

	return false;
}

bool ServiceConfig::get_uint(const rapidjson::Value& v, const std::string& key, unsigned short& i)
{
	if(v.HasMember(key.c_str()))
	{
		const rapidjson::Value& tmp = v[key.c_str()];
		if(tmp.IsUint())
		{
			i = tmp.GetUint();
			return i != 0;
		}
	}

	return false;
}

bool ServiceConfig::get_uint(const rapidjson::Value& v, const std::string& key, unsigned int& i)
{
	if(v.HasMember(key.c_str()))
	{
		const rapidjson::Value& tmp = v[key.c_str()];
		if(tmp.IsUint())
		{
			i = tmp.GetUint();
			return i != 0;
		}
	}

	return false;
}

bool ServiceConfig::get_bool(const rapidjson::Value& v, const std::string& key, bool& b)
{
	if(v.HasMember(key.c_str()))
	{
		const rapidjson::Value& tmp = v[key.c_str()];
		if(tmp.IsBool())
		{
			b = tmp.GetBool();
			return true;
		}
	}

	return false;
}

ServiceConfig::Service* ServiceConfig::get_service(unsigned short service_id)
{
	if(service_id != 0)
	{
		auto it = _service_map.find(service_id);
		if(it != _service_map.end())
		{
			return &it->second;
		}
	}

	return nullptr;
}

std::string ServiceConfig::get_service_list()
{
	 ServiceConfig sc = copy();
	 for(auto& it_service : sc._service_map)
	 {
		 Service& service = it_service.second;
		 service.depend_map.clear();
		 service.kv_map.clear();
		 service.heartbeat_list.clear();
		 service.inservice_list.clear();
	 }
	
	 return sc.map_to_json();
}

bool ServiceConfig::get_service_info(unsigned short service_id, std::string& json)
{
	auto p = get_service(service_id);
	if(p !=nullptr)
	{
		ServiceConfig sc;
		sc._service_map.insert(std::make_pair(p->service_id, *p));
		json = sc.map_to_json();
		return true;
	}

	return false;
}

bool ServiceConfig::set_service_heartbeat(unsigned short service_id, const Heartbeat& heartbeat)
{
	if(heartbeat.heartbeat_gap != 0 && heartbeat.lose_time != 0 && heartbeat.recover_time != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			p->heartbeat = heartbeat;
			return true;
		}
	}

	return false;
}

bool ServiceConfig::add_service_depend(unsigned short service_id, unsigned short depend_service_id)
{
	if(depend_service_id != 0)
	{
		auto p = get_service(service_id);
		auto p_depend_service_id = get_service(depend_service_id);
		if(p != nullptr && p_depend_service_id != nullptr)
		{
			auto it = p->depend_map.find(depend_service_id);
			if(it == p->depend_map.end())
			{
				Depend depend;
				depend.depend_service_id = depend_service_id;
				p->depend_map.insert(std::make_pair(depend.depend_service_id, depend));
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::del_service_depend(unsigned short service_id, unsigned short depend_service_id)
{
	if(depend_service_id != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it = p->depend_map.find(depend_service_id);
			if(it != p->depend_map.end())
			{
				p->depend_map.erase(it);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::set_service_kv(unsigned short service_id, const KV& kv)
{
	if(kv.key.size() > 0 && kv.val.size() > 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it = p->kv_map.find(kv.key);
			if(it != p->kv_map.end())
			{
				it->second = kv;
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::add_service_kv(unsigned short service_id, const KV& kv)
{
	if(kv.key.size() > 0 && kv.val.size() > 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it = p->kv_map.find(kv.key);
			if(it == p->kv_map.end())
			{
				p->kv_map.insert(std::make_pair(kv.key, kv));
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::del_service_kv(unsigned short service_id, const std::string& key)
{
	if(key.size() > 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it = p->kv_map.find(key);
			if(it != p->kv_map.end())
			{
				p->kv_map.erase(key);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::add_service_heartbeat_list(unsigned short service_id, const IPInfo& ip_info)
{
	if(ip_info.proc_id != 0 && ip_info.proc_des.size() > 0 && ip_info.in_ip.size() > 0 && ip_info.out_ip.size() > 0 && ip_info.port != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it_heartbeat_list = std::find(p->heartbeat_list.begin(), p->heartbeat_list.end(), ip_info.proc_id);
			auto it_inservice_list = std::find(p->inservice_list.begin(), p->inservice_list.end(), ip_info.proc_id);
			if(it_heartbeat_list == p->heartbeat_list.end() && it_inservice_list == p->inservice_list.end())
			{
				p->heartbeat_list.push_back(ip_info);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::del_service_heartbeat_list(unsigned short service_id, unsigned int proc_id)
{
	if(proc_id != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it = std::find(p->heartbeat_list.begin(), p->heartbeat_list.end(), proc_id);
			if(it != p->heartbeat_list.end())
			{
				p->heartbeat_list.erase(it);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::add_service_inservice_list(unsigned short service_id, unsigned int proc_id)
{
	if(proc_id != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it_heartbeat_list = std::find(p->heartbeat_list.begin(), p->heartbeat_list.end(), proc_id);
			auto it_inservice_list = std::find(p->inservice_list.begin(), p->inservice_list.end(), proc_id);
			if(it_heartbeat_list != p->heartbeat_list.end() && it_inservice_list ==  p->inservice_list.end())
			{
				IPInfo ip_info = *it_heartbeat_list;
				p->heartbeat_list.erase(it_heartbeat_list);
				p->inservice_list.push_back(ip_info);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::del_service_inservice_list(unsigned short service_id, unsigned int proc_id)
{
	if(proc_id != 0)
	{
		auto p = get_service(service_id);
		if(p != nullptr)
		{
			auto it_heartbeat_list = std::find(p->heartbeat_list.begin(), p->heartbeat_list.end(), proc_id);
			auto it_inservice_list = std::find(p->inservice_list.begin(), p->inservice_list.end(), proc_id);
			if(it_heartbeat_list == p->heartbeat_list.end() && it_inservice_list !=  p->inservice_list.end())
			{
				IPInfo ip_info = *it_inservice_list;
				p->inservice_list.erase(it_inservice_list);
				p->heartbeat_list.push_back(ip_info);
				return true;
			}
		}
	}

	return false;
}

bool ServiceConfig::add_service(unsigned short service_id, const std::string& service_name, const Heartbeat& heartbeat)
{
	if(service_id != 0 && service_name.size() > 0 && heartbeat.heartbeat_gap != 0 && heartbeat.lose_time != 0 && heartbeat.recover_time != 0)
	{
		for(auto& it_service : _service_map)
		{
			if(it_service.second.service_id == service_id || it_service.second.service_name == service_name)
			{
				return false;
			}
		}

		Service service;
		service.service_id = service_id;
		service.service_name = service_name;	
		service.heartbeat = heartbeat;
		_service_map.insert(std::make_pair(service.service_id, service));
		return true;
	}

	return false;
}

bool ServiceConfig::del_service(unsigned short service_id)
{
	if(service_id != 0)
	{
		auto it = _service_map.find(service_id);
		if(it != _service_map.end())
		{
			_service_map.erase(it);
			return true;
		}
	}

	return false;
}

void ServiceConfig::IPInfo::change()
{
	ip = Util::sip_2_uip(in_ip.c_str());
	/*
	memcpy((char*)&key_ip_port, (char*)&ip, sizeof(ip));
	memcpy((char*)&key_ip_port + sizeof(ip), (char*)&port, sizeof(port));
	*/
	key_ip_port = Util::uiui2ull(ip, port);
}

File: balancer/service/proxy/src/config/service_config.py
# -*- coding: utf-8 -*-

import json
from src.tool import util


class Heartbeat:
    def __init__(self):
        self.heartbeat_enable = False
        self.heartbeat_gap = 0
        self.lose_time = 0
        self.recover_time = 0

    def load(self, jo):
        self.heartbeat_enable = jo.heartbeat_enable
        self.heartbeat_gap = jo.heartbeat_gap
        self.lose_time = jo.lose_time
        self.recover_time = jo.recover_time

        heartbeat = Heartbeat()
        if type(heartbeat.heartbeat_enable) != type(self.heartbeat_enable):
            raise Exception('type failed')
        if type(heartbeat.heartbeat_gap) != type(self.heartbeat_gap) or \
                heartbeat.heartbeat_gap == self.heartbeat_gap:
            raise Exception('type failed')
        if type(heartbeat.lose_time) != type(self.lose_time) or \
                heartbeat.lose_time == self.lose_time:
            raise Exception('type failed')
        if type(heartbeat.recover_time) != type(self.recover_time) or \
                heartbeat.recover_time == self.recover_time:
            raise Exception('type failed')


class Depend:
    def __init__(self):
        self.depend_service_id = 0

    def load(self, jo):
        self.depend_service_id = jo.depend_service_id

        depend = Depend()
        if type(depend.depend_service_id) != type(self.depend_service_id) or \
                depend.depend_service_id == self.depend_service_id:
            raise Exception('type failed')


class KV:
    def __init__(self):
        self.key = ''
        self.val = ''

    def load(self, jo):
        self.key = jo.key
        self.val = jo.val

        kv = KV()
        if type(kv.key) != type(self.key) or kv.key == self.key:
            raise Exception('type failed')
        if type(kv.val) != type(self.val) or kv.val == self.val:
            raise Exception('type failed')


class IPInfo:
    def __init__(self):
        self.proc_id = 0
        self.proc_des = ''
        self.in_ip = ''
        self.in_port = 0
        self.out_ip = ''
        self.out_port = 0

    def load(self, jo):
        self.proc_id = jo.proc_id
        self.proc_des = jo.proc_des
        self.in_ip = jo.in_ip
        self.in_port = jo.in_port
        self.out_ip = jo.out_ip
        self.out_port = jo.out_port

        ip_info = IPInfo()
        if type(ip_info.proc_id) != type(self.proc_id) or ip_info.proc_id == self.proc_id:
            raise Exception('type failed')
        if type(ip_info.proc_des) != type(self.proc_des) or ip_info.proc_des == self.proc_des:
            raise Exception('type failed')
        if type(ip_info.in_ip) != type(self.in_ip) or ip_info.in_ip == self.in_ip:
            raise Exception('type failed')
        if type(ip_info.in_port) != type(self.in_port) or ip_info.in_port == self.in_port:
            raise Exception('type failed')
        if type(ip_info.out_ip) != type(self.out_ip) or ip_info.out_ip == self.out_ip:
            raise Exception('type failed')
        if type(ip_info.out_port) != type(self.out_port) or ip_info.out_port == self.out_port:
            raise Exception('type failed')


class Service:
    def __init__(self):
        self.service_id = 0
        self.service_name = ''
        self.heartbeat = Heartbeat()
        self.depend_map = {}
        self.kv_map = {}
        self.heartbeat_list = []
        self.inservice_list = []

    def load(self, jo):
        self.service_id = jo.service_id
        self.service_name = jo.service_name
        self.heartbeat.load(jo.heartbeat)
        for n in jo.depend_map:
            depend = Depend()
            depend.load(n)
            self.depend_map[depend.depend_service_id] = depend
        for n in jo.kv_map:
            kv = KV()
            kv.load(n)
            self.kv_map[kv.key] = kv
        for n in jo.heartbeat_list:
            ip_info = IPInfo()
            ip_info.load(n)
            self.heartbeat_list.append(ip_info)
        for n in jo.inservice_list:
            ip_info = IPInfo()
            ip_info.load(n)
            self.inservice_list.append(ip_info)

        service = Service()
        if type(service.service_id) != type(self.service_id) or service.service_id == self.service_id:
            raise Exception('type failed')
        if type(service.service_name) != type(self.service_name) or service.service_name == self.service_name:
            raise Exception('type failed')
        if type(service.heartbeat) != type(self.heartbeat):
            raise Exception('type failed')
        if type(service.depend_map) != type(self.depend_map):
            raise Exception('type failed')
        if type(service.kv_map) != type(self.kv_map):
            raise Exception('type failed')
        if type(service.heartbeat_list) != type(self.heartbeat_list):
            raise Exception('type failed')
        if type(service.inservice_list) != type(self.inservice_list):
            raise Exception('type failed')


class ServiceConfig:
    def __init__(self):
        self.service_map = {}
        self.json = ''

    def __str__(self):
        return self.json

    def json_to_map(self, str):
        jo = json.loads(str, object_hook=util.JsonObject)
        try:
            for n in jo.service_map:
                service = Service()
                service.load(n)
                self.service_map[service.service_id] = service
        except Exception as e:
            print(e)
            return False

        self.json = str
        return True

    def get_service(self, service_id):
        if service_id != 0:
            return self.service_map.get(service_id)

        return None


str = '{ "service_map": [ { "service_id": 10100, "service_name": "gate", "heartbeat": { "heartbeat_enable": false, "heartbeat_gap": 5, "lose_time": 3, "recover_time": 5 }, "depend_map": [ { "depend_service_id": 10200 } ], "kv_map": [ { "key": "timeout", "val": "30" }, { "key": "env", "val": "test" } ], "heartbeat_list": [ { "proc_id": 100, "proc_des": "gate_1_0", "in_ip": "121.1.1.1", "in_port": 10100, "out_ip": "11.1.1.1", "out_port": 30000 } ], "inservice_list": [ { "proc_id": 200, "proc_des": "gate_2_0", "in_ip": "121.1.1.2", "in_port": 10100, "out_ip": "11.1.1.2", "out_port": 30000 } ] }, { "service_id": 10200, "service_name": "group", "heartbeat": { "heartbeat_enable": true, "heartbeat_gap": 5, "lose_time": 3, "recover_time": 5 }, "depend_map": [ { "depend_service_id": 10100 } ], "kv_map": [ { "key": "name_max_size", "val": "1024" }, { "key": "buffer_max_size", "val": "1024" }, {  "key": "navigate_key",  "val": "qsdjkajwojfdowjowj"  },  {  "key": "navigate_key_timeout_us",  "val": "60000000"  },  { "key": "map_gate_service_id",  "val": "10300" } ], "heartbeat_list": [ { "proc_id": 100, "proc_des": "group_1_0", "in_ip": "121.1.1.10", "in_port": 10200, "out_ip": "11.1.1.10", "out_port": 10200 } ], "inservice_list": [ { "proc_id": 200, "proc_des": "group_2_0", "in_ip": "121.1.1.20", "in_port": 10200, "out_ip": "11.1.1.20", "out_port": 10200 }, { "proc_id": 201, "proc_des": "group_2_1", "in_ip": "121.1.1.21", "in_port": 10201, "out_ip": "11.1.1.21", "out_port": 10201 } ] } ] } '

sc = ServiceConfig()
sc.json_to_map(str)

print(sc)





